File: backend/BoundfoxStudios.Smudgy.Services/SessionService.cs
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using BoundfoxStudios.Smudgy.Data;
using BoundfoxStudios.Smudgy.Data.DTOs;
using BoundfoxStudios.Smudgy.Data.Entities;
using BoundfoxStudios.Smudgy.Data.Models;
using Microsoft.AspNetCore.SignalR;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace BoundfoxStudios.Smudgy.Services
{
  public class SessionService
  {
    private readonly IGroupManager _groupManager;
    private readonly ILogger<SessionService> _logger;
    private readonly SmudgyContext _context;
    private readonly IHubClients _clients;

    public SessionService(
      IGroupManager groupManager,
      IHubClients clients,
      ILogger<SessionService> logger,
      SmudgyContext context
    )
    {
      _groupManager = groupManager;
      _logger = logger;
      _context = context;
      _clients = clients;
    }

    public async Task<Guid> CreateSessionAsync(SessionConfiguration configuration, string hostPlayerConnectionId,
      CancellationToken cancellationToken = default)
    {
      _logger.LogInformation("Creating new session for {PlayerId}", hostPlayerConnectionId);

      var hostPlayer = await _context.Players.SingleOrDefaultAsync(p => p.SocketId == hostPlayerConnectionId, cancellationToken);

      if (hostPlayer == null)
      {
        // TODO: Error handling
        throw new Exception("host player not found");
      }

      var session = new Session()
      {
        Id = Guid.NewGuid(),
        HostPlayerId = hostPlayer.Id,
        Language = configuration.Language,
        State = SessionState.Lobby,
        MaxPlayers = configuration.MaxPlayers,
        RoundsToPlay = configuration.RoundsToPlay,
        RoundTimeInSeconds = configuration.RoundTimeInSeconds
      };

      await _context.Sessions.AddAsync(session, cancellationToken);
      await _context.SaveChangesAsync(cancellationToken);

      _logger.LogInformation("Created new session {id} for {PlayerId}", session.Id, hostPlayerConnectionId);

      return session.Id;
    }

    public async Task<SessionConfiguration> JoinSessionAsync(string playerId, Guid sessionId, CancellationToken cancellationToken = default)
    {
      _logger.LogInformation("Player {Player} wants to join session {Session}", playerId, sessionId);

      var session = await _context.Sessions.Include(p => p.Players).SingleOrDefaultAsync(p => p.Id == sessionId, cancellationToken);
      var player = await _context.Players.SingleOrDefaultAsync(p => p.SocketId == playerId, cancellationToken);

      if (session == null || player == null)
      {
        _logger.LogInformation("Player {Player} can not join session {Session}, either player or session was not found", playerId,
          sessionId);
        return null;
      }

      if (session.Players.All(p => p.Id != player.Id))
      {
        session.Players.Add(player);
        await _context.SaveChangesAsync(cancellationToken);
      }

      await _groupManager.AddToGroupAsync(playerId, session.Id.ToString(), cancellationToken);
      await _clients.Group(session.Id.ToString()).SendAsync("playerJoinSession",
        new PlayerListItem() { Id = player.Id, Name = player.Name }, cancellationToken);

      return new SessionConfiguration()
      {
        RoundsToPlay = session.RoundsToPlay,
        Language = session.Language,
        MaxPlayers = session.MaxPlayers,
        RoundTimeInSeconds = session.RoundTimeInSeconds
      };
    }

    public async Task<PlayerListItem[]> PlayerListAsync(Guid sessionId, CancellationToken cancellationToken = default)
    {
      return await _context.Sessions.Where(p => p.Id == sessionId).SelectMany(p => p.Players)
        .Select(p => new PlayerListItem() { Id = p.Id, Name = p.Name })
        .ToArrayAsync(cancellationToken);
    }

    public async Task UpdateSessionConfigurationAsync(Guid sessionId, string playerId, SessionConfiguration sessionConfiguration,
      CancellationToken cancellationToken = default)
    {
      _logger.LogInformation("Receiving new session configuration from {Player} for {Session}: {@SessionConfiguration}", playerId,
        sessionId,
        sessionConfiguration);

      var player = await _context.Players.SingleOrDefaultAsync(p => p.SocketId == playerId, cancellationToken);
      var session = await _context.Sessions.SingleOrDefaultAsync(p => p.HostPlayerId == player.Id && p.Id == sessionId, cancellationToken);

      if (session == null)
      {
        _logger.LogWarning("Session {Session} not found for player {Player}", sessionId, player.Id);
        return;
      }

      session.Language = sessionConfiguration.Language;
      session.RoundsToPlay = sessionConfiguration.RoundsToPlay;
      session.RoundTimeInSeconds = sessionConfiguration.RoundTimeInSeconds;
      session.MaxPlayers = sessionConfiguration.MaxPlayers;

      await _context.SaveChangesAsync(cancellationToken);

      await _clients.Group(sessionId.ToString()).SendAsync("updateSessionConfiguration", new SessionConfiguration()
      {
        Language = session.Language,
        MaxPlayers = session.MaxPlayers,
        RoundsToPlay = session.RoundsToPlay,
        RoundTimeInSeconds = session.RoundTimeInSeconds
      }, cancellationToken);
    }

    public async Task<bool> StartGameAsync(Guid sessionId, string playerId, CancellationToken cancellationToken)
    {
      _logger.LogInformation("Receiving start game request from {Player} for {Session}.", playerId, sessionId);

      var player = await _context.Players.SingleOrDefaultAsync(p => p.SocketId == playerId, cancellationToken);
      var session = await _context.Sessions.SingleOrDefaultAsync(p => p.HostPlayerId == player.Id && p.Id == sessionId, cancellationToken);

      if (session == null)
      {
        _logger.LogWarning("Session {Session} not found for player {Player}", sessionId, player.Id);
        return false;
      }

      if (session.State != SessionState.Lobby)
      {
        _logger.LogWarning("Session {Session} is not in state lobby, but {State}. Can not start anymore.", sessionId, session.State);
        return false;
      }

      session.State = SessionState.InGame;

      await _context.SaveChangesAsync(cancellationToken);

      await _clients.Group(sessionId.ToString()).SendAsync("startGame", cancellationToken);

      return true;
    }
  }
}

File: backend/src/modules/session/services/session.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { In, Repository } from 'typeorm';
import { Guid } from '../../../models/guid';
import { newUuid } from '../../../utils/uuid';
import { Player } from '../../player/entities/player.entity';
import { PlayerService } from '../../player/services/player.service';
import { Session, SessionEntitySchema } from '../entities/session.entity';
import { PlayerListItem } from '../models/player-list-item';
import { SessionConfiguration } from '../models/session-configuration';
import { SessionState } from '../models/session-state';

@Injectable()
export class SessionService {
  private readonly logger = new Logger(SessionService.name);

  constructor(
    @InjectRepository(SessionEntitySchema)
    private readonly sessionEntitiesRepository: Repository<Session>,
    private readonly playerService: PlayerService,
  ) {}

  async createSession(configuration: SessionConfiguration, hostPlayerSocketId: string): Promise<Guid> {
    this.logger.log(`Creating new session for ${hostPlayerSocketId}`);

    const hostPlayer = await this.playerService.findBySocketId(hostPlayerSocketId);

    if (!hostPlayer) {
      // TODO: Error handling
      throw new Error(`Host player with socket id ${hostPlayerSocketId} not found`);
    }

    const session = this.sessionEntitiesRepository.create({
      id: newUuid(),
      hostPlayer,
      language: configuration.language,
      maxPlayers: configuration.maxPlayers,
      roundsToPlay: configuration.roundsToPlay,
      roundTimeInSeconds: configuration.roundTimeInSeconds,
      sessionState: SessionState.Lobby,
      players: [],
    });

    const { id } = await this.sessionEntitiesRepository.save(session);

    this.logger.log(`Created new session ${id} for ${hostPlayer.id}`);

    return id;
  }

  async joinSession(sessionId: Guid, socketId: string): Promise<{ session: Session; player: Player } | undefined> {
    this.logger.log(`PlayerSocketId ${socketId} wants to join session ${sessionId}`);

    const session = await this.sessionEntitiesRepository.findOne({
      where: {
        id: sessionId,
      },
      relations: {
        players: true,
      },
    });

    const player = await this.playerService.findBySocketId(socketId);

    if (!session || !player) {
      this.logger.warn(`PlayerId ${socketId} can not join session ${sessionId}. Either player or session was not found.`);
      return;
    }

    if (session.players && session.players.every(player => player.id !== socketId)) {
      session.players.push(player);
      await this.sessionEntitiesRepository.save(session);
    }

    return { session, player };
  }

  async playerList(sessionId: Guid): Promise<PlayerListItem[]> {
    const session = await this.sessionEntitiesRepository.findOne({
      where: {
        id: sessionId,
      },
      relations: {
        players: true,
      },
    });

    return (session?.players ?? []).map(player => ({ id: player.id, name: player.name }));
  }

  async disconnectPlayerFromAllSessions(socketId: string): Promise<{ sessions: Guid[]; player: PlayerListItem } | undefined> {
    const player = await this.playerService.findBySocketId(socketId);

    if (!player) {
      this.logger.warn(`PlayerId ${socketId} was not found.`);
      return;
    }

    const sessions = await this.findSessionWherePlayerParticipates(player);

    for (const session of sessions) {
      session.players = (session.players ?? []).filter(sessionPlayer => sessionPlayer.id !== player.id);
    }

    await this.sessionEntitiesRepository.save(sessions);

    return {
      player: {
        name: player.name,
        id: player.id,
      },
      sessions: sessions.map(session => session.id),
    };
  }

  // TODO: Change from Promise<bool> to <void> and throw an error when something is not right here.
  async updateSessionConfiguration(sessionId: Guid, socketId: string, sessionConfiguration: SessionConfiguration): Promise<boolean> {
    this.logger.log(`Receiving new session configuration from ${socketId} for ${sessionId}: ${JSON.stringify(sessionConfiguration)}`);

    const session = await this.sessionEntitiesRepository.findOne({
      where: { id: sessionId },
      relations: { hostPlayer: true },
    });

    if (!session) {
      this.logger.log(`Did not find session with session id ${sessionId}`);
      return false;
    }

    if (session.hostPlayer.socketId !== socketId) {
      this.logger.log(`Player (SocketId ${socketId})) tried to change configuration where it is not host player`);
      return false;
    }

    session.language = sessionConfiguration.language;
    session.maxPlayers = sessionConfiguration.maxPlayers;
    session.roundTimeInSeconds = sessionConfiguration.roundTimeInSeconds;
    session.roundsToPlay = sessionConfiguration.roundsToPlay;

    await this.sessionEntitiesRepository.save(session);

    return true;
  }

  // TODO: Change from Promise<bool> to <void> and throw an error when something is not right here.
  async startGame(sessionId: Guid, socketId: string): Promise<boolean> {
    this.logger.log(`Receiving start game request from ${socketId} for ${sessionId}`);

    const session = await this.sessionEntitiesRepository.findOne({
      where: { id: sessionId },
      relations: { hostPlayer: true },
    });

    if (!session) {
      this.logger.warn(`Session ${sessionId} not found.`);
      return false;
    }

    if (session.hostPlayer.socketId !== socketId) {
      this.logger.warn(`Player (SocketId ${socketId}) tried to start the game, but its not the host player`);
      return false;
    }

    if (session.sessionState !== SessionState.Lobby) {
      this.logger.warn(`Session ${sessionId} is not in state ${SessionState.Lobby}`);
      return false;
    }

    session.sessionState = SessionState.InGame;

    await this.sessionEntitiesRepository.save(session);

    return true;
  }

  async findSessionWherePlayerParticipates(player: Player): Promise<Session[]> {
    const sessionState = In([SessionState.Lobby, SessionState.InGame]);
    const sessions = await this.sessionEntitiesRepository.find({
      select: { id: true },
      where: { players: { id: player.id }, sessionState },
    });

    const sessionIds = sessions.map(session => session.id);

    return this.sessionEntitiesRepository.find({ where: { id: In(sessionIds) }, relations: { players: true } });
  }
}
